docs/4.5. Behaviour - Resource Servers.md

```diff
@@ -4,19 +4,29 @@ _(c) AMWA 2019, CC Attribution-NoDerivatives 4.0 International (CC BY-ND 4.0)_
 
 ## Public keys
 
-Resource Servers SHOULD seek to fetch public keys from the Authorization Server at least once every hour, altering
-their retrieval time by a value between 0-60 seconds to avoid overloading the Authorization Server due to Resource
-Servers synchronising their retrieval time.
+Resource Servers SHOULD seek to fetch public keys from one Authorization Server in the deployment at least once every
+hour, altering their retrieval time by a value between 0-60 seconds to avoid overloading the Authorization Server due to
+Resource Servers synchronising their retrieval time.
 
-If a Resource Server is unable to contact the Authorization Server, the Resource Server MUST implement a random
-back-off mechanism to avoid overloading the Authorization Server in the event of a system restart. Also if a
-Resource Server is unable to contact an Authorization Server, the Resource Server MAY assume currently held public
-keys remain valid until it is able to re-establish a connection to an Authorization Server.
+If a Resource Server is unable to contact an Authorization Server, the Resource Server SHOULD attempt to contact another
+Authorization Server from the discovered list until this either succeeds or the list is exhausted. If no Authorization
+Servers respond as expected the Resource Server MUST implement a randomised back-off mechanism to avoid overloading the
+Authorization Servers.
+
+If a Resource Server is unable to contact an Authorization Server, the Resource Server SHOULD assume currently held
+public keys remain valid until it is able to re-establish a connection to an Authorization Server.
 
 Resource Servers SHOULD attempt to verify tokens against all keys presented at the Authorization Server's public key
 endpoint. All valid JWK's SHOULD be tried until the token is verified or until no keys are left.
 
-If the Resource Server fails to verify all public keys available, it MUST reject the token.
+Where a Resource Server has no matching public key for a given token, it SHOULD attempt to obtain the missing public key
+via the the token `iss` claim. Resource Servers SHOULD use the hostname and port from the issuer claim in combination
+with any matching DNS-SD TXT records when it is necessary to obtain a missing public key.
+
```

Comment:
Does this conflict with the documentation in [Discovery](https://github.com/AMWA-TV/nmos-authorization/blob/v1.0-dev/docs/3.0.%20Discovery.md)?

In our understanding the `iss` claim should be a URL like `<api_proto>://<hostname>:<port>[/<api_selector>]`. That URL includes all the same info from the DNS-SD records, right? Then, the Resource Server could fetch the server metadata from `<api_proto>://<hostname>:<port>/.well-known/oauth-authorization-server[/<api_selector>]` in order to discover the `jwks_uri` to fetch new public keys.

Is it required that a Resource Server to be able to cache information (public keys and so on) from multiple Auth Servers?

And what if the details from the `iss` URL are for a bad actor (or simply one the Resource Server hasn't discovered or been configured with)?

Reply:
It would be nice to clarify whether the current request MAY/SHOULD/MUST be rejected with 401 Unauthorized in this case too. (Blocking the request while the Resource Server attempts to obtain the missing public key seems like a bad idea!)

Reply:
The OAuth and JWT specs only require a StringOrURI for the issuer. However, it looks like a side effect of specifying RFC 8414 is that this needs to be a URL based upon the first sentence of https://tools.ietf.org/html/rfc8414#section-3. I can certainly tighten up the 'iss' language and simplify this based upon that.

If the `iss` is from a bad actor yet they form part of the same domain as a valid auth server and they have a certificate which is trusted according to the CA cert held by the device then I don't think there's much we can do - the system is already compromised.

Rather than a 401 if we're trying to prevent blocking I wonder whether a 503 might be more appropriate? You wouldn't want the client to go off and try to get a new token because of a temporary issue.

Reply:
> Is it required that a Resource Server to be able to cache information (public keys and so on) from multiple Auth Servers?

Caching is optional... SHOULD? Since the main goal is to improve performance. I have not seen any implementation which have support for multiple auth servers and caching JWKS. :thinking: There's no mention of this in the draft https://tools.ietf.org/html/rfc6819

> And what if the details from the iss URL are for a bad actor (or simply one the Resource Server hasn't discovered or been configured with)?

When a token fails to be validated, you should be refreshing the cached data and retry. I would say that includes discovery.

> It would be nice to clarify whether the current request MAY/SHOULD/MUST be rejected with 401 Unauthorized in this case too. (Blocking the request while the Resource Server attempts to obtain the missing public key seems like a bad idea!)

A 401 may trick the client to obtain a new token, maybe a [503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) but would that be too complicated?
I don't think we can require a non-blocking behavior since it's not a common implementation. ( have not seen it in the common OSS libraries )

The keys are not expected to change for long period of time. One request every 4,6, or 12 months might get delayed from a stale cache entry. Is that acceptable in our industry?

Reply:
> iss: The contents of this claim SHOULD match one entry in the common name field or alternate names fields in all TLS certificates used by the Authorization Server

Doing so, is that mean the resource server will have to communicate to each of the discovered authorization servers to obtain their certificates for their CN and SANs.

> In cases where the Resource Server needs to fetch a public key from a remote Authorization Server it MAY temporarily respond with an HTTP 503 code in order to avoid blocking the incoming authorized request.

In event of using 503 to trigger refetching public keys, should response also contain the Retry-After: HTTP header.

Reply:
Just to clarify my concern about the token's iss claim, I am looking into how the this claim can be verified.  Maybe this is not required, the iss should only be used for refetching missing public key.